metron-analytics/metron-profiler-common/src/main/java/org/apache/metron/profiler/hbase/DecodableRowKeyBuilder.java

```diff
@@ -0,0 +1,402 @@
+    return ProfilePeriod.visitPeriods( start
+            , end
+            , periodDurationMillis
+            , TimeUnit.MILLISECONDS
+            , Optional.empty()
+            , period -> encode(profile, entity, groups, period)
+    );
+
+  }
+
+  /**
+   * Builds a list of row keys necessary to retrieve a profile's measurements over
+   * a time horizon.
+   * <p>
+   * This method is useful when attempting to read ProfileMeasurements stored in HBase.
+   *
+   * @param profile    The name of the profile.
+   * @param entity     The name of the entity.
+   * @param groups     The group(s) used to sort the profile data.
+   * @param periods    The profile measurement periods to compute the rowkeys for
+   * @return All of the row keys necessary to retrieve the profile measurements.
+   */
+  @Override
+  public List<byte[]> encode(String profile, String entity, List<Object> groups, Iterable<ProfilePeriod> periods) {
+    List<byte[]> rowKeys = new ArrayList<>();
+    for(ProfilePeriod period : periods) {
+      rowKeys.add(encode(profile, entity, groups, period));
+    }
+    return rowKeys;
+  }
+
+  /**
+   * Builds the row key for a given profile measurement.
+   * @param m The profile measurement.
+   * @return The HBase row key.
+   */
+  @Override
+  public byte[] encode(ProfileMeasurement m) {
+    return encode(m.getProfileName(), m.getEntity(), m.getGroups(), m.getPeriod());
+  }
+
+  /**
+   * Build the row key.
+   * @param profile The name of the profile.
+   * @param entity The name of the entity.
+   * @param period The period in which the measurement was taken.
+   * @param groups The groups.
+   * @return The HBase row key.
+   */
+  public byte[] encode(String profile, String entity, List<Object> groups, ProfilePeriod period) {
+
+    if(profile == null)
+      throw new IllegalArgumentException("Cannot encode row key; invalid profile name.");
+    if(entity == null)
+      throw new IllegalArgumentException("Cannot encode row key; invalid entity name.");
+    if(period == null)
+      throw new IllegalArgumentException("Cannot encode row key; invalid profile period.");
+
+    long periodId = period.getPeriod();
+    long periodDurationMillis = period.getDurationMillis();
+
+    byte[] salt = encodeSalt(periodId, saltDivisor);
+    byte[] profileB = Bytes.toBytes(profile);
+    byte[] entityB = Bytes.toBytes(entity);
+    byte[] groupB = encodeGroups(groups);
+
+    int capacity = Short.BYTES + 1 + salt.length + profileB.length + entityB.length + groupB.length + (Integer.BYTES * 3) + (Long.BYTES * 2);
+    ByteBuffer buffer = ByteBuffer
+            .allocate(capacity)
+            .order(byteOrder)
+            .putShort(MAGIC_NUMBER)
+            .put(VERSION)
+            .putInt(salt.length)
+            .put(salt)
+            .putInt(profileB.length)
+            .put(profileB)
+            .putInt(entityB.length)
+            .put(entityB)
+            .put(groupB)
```

Comment:
@mattf-horton A 'length' the number of groups along with the size of each group name is encoded.  See the function `encodeGroup`.

@cestella I would be fine with shorts instead of ints.